common/authorization/rsaTokenKeyProvider.go

```diff
@@ -0,0 +1,164 @@
+// THE SOFTWARE.
+
+package authorization
+
+import (
+	"crypto/ecdsa"
+	"crypto/rsa"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/dgrijalva/jwt-go"
+
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/loggerimpl"
+	"go.temporal.io/server/common/log/tag"
+	"go.temporal.io/server/common/service/config"
+)
+
+type jwks struct {
+	Keys []jsonWebKeys `json:"keys"`
+}
+
+type jsonWebKeys struct {
+	Kty string   `json:"kty"`
+	Kid string   `json:"kid"`
+	Use string   `json:"use"`
+	N   string   `json:"n"`
+	E   string   `json:"e"`
+	X5c []string `json:"x5c"`
+}
+
+// Default RSA token key provider
+type rsaKeyProvider struct {
+	config   config.JWTKeyProvider
+	keys     map[string]*rsa.PublicKey
+	keysLock sync.RWMutex
+	ticker   *time.Ticker
+	logger   log.Logger
+	stop     chan bool
+}
+
+var _ TokenKeyProvider = (*rsaKeyProvider)(nil)
+
+func NewRSAKeyProvider(cfg *config.Config) *rsaKeyProvider {
+	logger := loggerimpl.NewLogger(cfg.Log.NewZapLogger())
+	provider := rsaKeyProvider{config: cfg.Global.Authorization.JWTKeyProvider, logger: logger}
+	provider.init()
+	return &provider
+}
+
+func (a *rsaKeyProvider) init() {
+	a.keys = make(map[string]*rsa.PublicKey)
+	if len(a.config.KeySourceURIs) > 0 {
+		err := a.updateKeys()
+		if err != nil {
+			a.logger.Error("error during initial retrieval of RSA token keys: ", tag.Error(err))
+		}
+	}
+	if a.config.RefreshInterval > 0 {
+		a.stop = make(chan bool)
+		a.ticker = time.NewTicker(a.config.RefreshInterval)
+		go a.timerCallback()
+	}
+}
+
+func (a *rsaKeyProvider) rsaKey(alg string, kid string) (*rsa.PublicKey, error) {
+	if !strings.EqualFold(alg, "rs256") {
+		return nil, fmt.Errorf("unexpected signing algorithm: %s", alg)
+	}
+
+	a.keysLock.RLock()
+	key, found := a.keys[kid]
+	a.keysLock.RUnlock()
+	if !found {
+		return nil, fmt.Errorf("RSA key not found for key ID: %s", kid)
+	}
+	return key, nil
+}
+
+func (a *rsaKeyProvider) timerCallback() {
+	for {
+		select {
+		case <-a.stop:
+			break
+		case <-a.ticker.C:
+		}
+		if len(a.config.KeySourceURIs) > 0 {
+			err := a.updateKeys()
+			if err != nil {
+				a.logger.Error("error while refreshing RSA token keys: ", tag.Error(err))
+			}
+		}
+	}
+}
+
+func (a *rsaKeyProvider) updateKeys() error {
+	if len(a.config.KeySourceURIs) == 0 {
+		return fmt.Errorf("no URIs configured for retrieving keys")
+	}
+
+	keys := make(map[string]*rsa.PublicKey)
+
+	for _, uri := range a.config.KeySourceURIs {
+		resp, err := http.Get(uri)
+		if err != nil {
+			return err
+		}
+		defer resp.Body.Close()
+
+		jwks := jwks{}
+		err = json.NewDecoder(resp.Body).Decode(&jwks)
+		if err != nil {
+			return err
+		}
+
+		for _, k := range jwks.Keys {
+			cert := "-----BEGIN CERTIFICATE-----\n" + k.X5c[0] + "\n-----END CERTIFICATE-----"
+			key, err := jwt.ParseRSAPublicKeyFromPEM([]byte(cert))
+			if err != nil {
+				return err
+			}
+			keys[k.Kid] = key
+		}
```

Comment:
plz make this section a function
one more thing, the `defer resp.Body.Close()` will not be executed until `updateKeys` returns